### PR TITLE
channel notification - jde

### PR DESCRIPTION
**Pull request by johandem**
task_id: 1859802

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mail/models/mail_thread.py [hunk 1]
```diff
@@ -2027,7 +2027,7 @@ def message_post(self, body='', subject=None,
             'parent_id': parent_id,
             'subtype_id': subtype_id,
             'partner_ids': [(4, pid) for pid in partner_ids],
-            'channel_ids': kwargs.get('channel_ids', []),
+            'channel_ids': [(4, cid) for cid in kwargs.get('channel_ids', [])],
```
Comment (tde-banana-odoo) [inline addons/mail/models/mail_thread.py]: Actually most of message_post arguments should be like a new message creation values. Therefore either you give a valid orm command to message_post in JS, either you process them before to correctly support either a list of ids, either a list of commands. Code here is not correct as you break valid behavior for methods already giving a valid ORM command for channel_ids .

File: addons/mail/static/src/js/composers/basic_composer.js [hunk 2]
```diff
@@ -402,7 +402,7 @@ var BasicComposer = Widget.extend({
         // behavior for the chatter composer
 
         //Removing unwanted extra spaces from message
-        var value = _.escape(this.$input.val()).trim();
+        var value = this.$input.val().trim();
```
Comment (tde-banana-odoo) [inline addons/mail/static/src/js/composers/basic_composer.js]: Why this change ?

File: addons/mail/static/src/js/composers/mention_manager.js [hunk 3]
```diff
@@ -142,22 +142,23 @@ var MentionManager = Widget.extend({
         var mentionLink = "<a " + linkParams + " >" +
                                 "%s%s" +
                             "</a>";
-        _.each(this._listeners, function (listener) {
+        _.each(this._listeners, function (listener) {
```
Comment (tde-banana-odoo) [inline addons/mail/static/src/js/composers/mention_manager.js]: HumanLinter: extra tab ?

File: addons/mail/static/src/js/discuss.js [hunk 4]
```diff
@@ -642,12 +642,12 @@ var Discuss = AbstractAction.extend({
         if (type === 'multi_user_channel') {
             $input.autocomplete({
                 source: function (request, response) {
-                    self._lastSearchVal = _.escape(request.term);
+                    self._lastSearchVal = request.term;
```
Comment (tde-banana-odoo) [inline addons/mail/static/src/js/discuss.js]: Is it really safe ?

File: addons/mail/static/src/js/discuss.js [hunk 5]
```diff
@@ -647,7 +647,7 @@ var Discuss = AbstractAction.extend({
                         result.push({
                             label:  _.str.sprintf(
                                         '<strong>' + _t("Create %s (Public)") + '</strong>',
-                                        '<em>"#' + self._lastSearchVal + '"</em>'
+                                        '<em>"#' + _.escape(self._lastSearchVal) + '"</em>'
```
Comment (alexkuhn) [inline addons/mail/static/src/js/discuss.js]: no double escaping, drop this `_.escape`

File: addons/mail/static/src/js/composers/mention_manager.js [hunk 6]
```diff
@@ -354,7 +354,7 @@ var MentionManager = Widget.extend({
     _validateSelection: function (selection, inputMentions) {
         var validatedSelection = [];
         _.each(inputMentions, function (mention) {
-            var validatedMention = _.findWhere(selection, { name: mention.slice(1) });
+            var validatedMention = _.findWhere(selection, { name: _.escape(mention.slice(1)) });
```
Comment (alexkuhn) [inline addons/mail/static/src/js/composers/mention_manager.js]: remove this `_.escape`, shouldn't be there

File: addons/mail/static/src/js/composers/mention_manager.js [hunk 7]
```diff
@@ -157,7 +157,7 @@ var MentionManager = Widget.extend({
                     var selectionID = self._getSelectionID(match, selection);
                     // put back white spaces instead of non-breaking spaces in
                     // mention's name
-                    var matchName = match[0].substring(1)
+                    var matchName = _.unescape(match[0]).substring(1)
```
Comment (alexkuhn) [inline addons/mail/static/src/js/composers/mention_manager.js]: remove this `_.unescape`, shouldn't be there

File: addons/mail/static/src/js/composers/basic_composer.js [hunk 8]
```diff
@@ -416,6 +416,7 @@ var BasicComposer = Widget.extend({
             attachment_ids: _.pluck(this.get('attachment_ids'), 'id'),
             partner_ids: _.uniq(_.pluck(this._mentionManager.getListenerSelection('@'), 'id')),
             canned_response_ids: _.uniq(_.pluck(this._mentionManager.getListenerSelections()[':'], 'id')),
+            channel_ids: _.uniq(_.pluck(this._mentionManager.getListenerSelection('#'), 'id')),
```
Comment (alexkuhn) [inline addons/mail/static/src/js/composers/basic_composer.js]: update `getListenerSelection` so that it escapes the composer input val:
```js
...
var escapedVal = _.escape(thiis._composer.$input.val());
var inputMentions = escapedVal.match(new RegExp(delimiter+'[^ ]+(?= |&nbsp;)', 'g'));
...
```

File: addons/mail/static/src/js/models/threads/thread.js [hunk 9]
```diff
@@ -342,8 +342,13 @@ var Thread = AbstractThread.extend(ServicesMixin, {
         // the current design makes this quite hard to do.
         var body = mailUtils.parseAndTransform(_.str.trim(data.content), mailUtils.addLink);
         body = this._generateEmojis(body);
+        var channel_ids = [];
```
Comment (alexkuhn) [inline addons/mail/static/src/js/models/threads/thread.js]: camelCase, `channel_ids` => `channelIDs`

File: addons/mail/static/src/js/models/threads/thread.js [hunk 10]
```diff
@@ -342,8 +342,13 @@ var Thread = AbstractThread.extend(ServicesMixin, {
...
+        for (var i = 0; i < data.channel_ids.length; i++) {
+            channel_ids[i] = [4, data.channel_ids[i], false];
+        }
         var messageData = {
             partner_ids: data.partner_ids,
+            channel_ids: channel_ids,
```
Comment (alexkuhn) [inline addons/mail/static/src/js/models/threads/thread.js]:
```js
channel_ids: _.map(data.channel_ids, function (channelID) {
   return [4, channelID, false];
}),
```

File: addons/mail/static/src/js/composers/basic_composer.js (hunk 8 above)
Comment (alexkuhn) [reply addons/mail/static/src/js/composers/basic_composer.js]: this should fix an issue in 12.0, in which partners with special characters do not receive needaction when mentioned

File: addons/mail/static/src/js/composers/mention_manager.js [hunk 11]
```diff
@@ -315,7 +313,9 @@ var MentionManager = Widget.extend({
         }
         if (suggestions.length) {
             this.$el.html(QWeb.render(this._activeListener.suggestionTemplate, {
-                suggestions: suggestions,
+                suggestions: _.map(suggestions, function (suggestion) {
+                   return {'id': suggestion['id'], 'public': suggestion['public'], 'name': _.unescape(suggestion['name'])};
+                })
```
Comment (dbeguin) [inline addons/mail/static/src/js/composers/mention_manager.js]: unescaping only for UI rendering (avoid having this : "Project &amp; others" but having this instead : "Project & others")

File: addons/mail/static/src/js/composers/mention_manager.js [hunk 12]
```diff
@@ -385,7 +385,7 @@ var MentionManager = Widget.extend({
             // no substitution string given, so use the mention name instead
             // replace white spaces with non-breaking spaces to facilitate
             // mentions detection in text
-            selectedSuggestion.name = selectedSuggestion.name.replace(/ /g, NON_BREAKING_SPACE);
+            selectedSuggestion.name = _.unescape(selectedSuggestion.name.replace(/ /g, NON_BREAKING_SPACE));
```
Comment (dbeguin) [inline addons/mail/static/src/js/composers/mention_manager.js]: unescaping only for UI rendering (avoid having this : "Project &amp; others" but having this instead : "Project & others")

File: addons/mail/static/src/js/composers/mention_manager.js (hunk 11 above)
Comment (alexkuhn) [reply addons/mail/static/src/js/composers/mention_manager.js]:
```js
// mention manager stores escaped suggestion names
return _.extend({}, suggestion, { name: _.unescape(suggestion.name) });
```

File: addons/mail/static/src/js/composers/mention_manager.js (hunk 12 above)
Comment (alexkuhn) [reply addons/mail/static/src/js/composers/mention_manager.js]: put following comment before this LOC:
```js
// mention manager stores escaped suggestion names
```

**Comment by dbeguin** [comment]
@robodoo r+
